### PR TITLE
feat: Alow mapping lua functions

### PR DESCRIPTION
**Pull request by shadmansaleh**
The things I was talking about in [reddit](https://www.reddit.com/r/neovim/comments/ofncvw/comment/h4dof59)

It's not completely polished . You can take a look .

Edit: Since you've already removed map() call . I've removed related chages . Now this only has lua function mapping .

### CODE REVIEW

**Comment by Iron-E** [comment]
Thank you for drafting this! I will review and test it shortly.

File: lua/cartographer.lua [hunk 1]
```diff
@@ -1,6 +1,25 @@
 --- The Neovim API
 local api = vim.api
 
+-- See if this script is executing firsttime
+local ok, callbacks = pcall(require, 'cartographer.callbacks')
```
Comment (Iron-E) [inline lua/cartographer.lua]: We can add a new file for this so we don't have to `pcall`, `lua/cartographer/callbacks.lua`. We can put `callbacks.new` there too

File: lua/cartographer.lua [hunk 2]
```diff
@@ -50,6 +69,12 @@ local MetaCartographer =
 				unique = rawget(self, 'unique'),
 			}
 
+			if type(rhs) == 'function' then
+				local id = callbacks.new(rhs)
+				rhs = '<cmd>lua require("cartographer.callbacks")['..tostring(id)..']()<cr>'
```
Comment (Iron-E) [inline lua/cartographer.lua]: We don't need `tostring` here, since Lua can concat numbers

**Review by Iron-E — CHANGES_REQUESTED** [review]
I want to move that `cartographer.callbacks` out into another file so we can just `require` it without using `pcall`, and other than that I think the solution is really clever!

**Comment by Iron-E** [comment]
It works! Awesome. I'll document it and then merge, if you're satisfied as well :+1:

File: lua/cartographer.lua (hunk 2 above)
Comment (shadmansaleh) [reply lua/cartographer.lua]: That's news to me . I thought it was an error to concat numbers and string in lua 😂

**Comment by shadmansaleh** [comment]
It's ok you can merge . Though if someone reloads the callback file the callbacks will be lost . There's really no reason for someone to do that so 👍🏽

**Comment by Iron-E** [comment]
That's a good point, though like you said one really has to go out of their way to clear it.

When testing too, I noticed each new test would increment which `id` was being used, despite the function already being in the table. In the future if it causes performance issues I might have to track the functions another way, but I don't think fixing it will be any less of a performance loss.

Thanks again for all the help!

**Comment by shadmansaleh** [comment]
> When testing too, I noticed each new test would increment which id was being used, despite the function already being in the table. In the future if it causes performance issues I might have to track the functions another way, but I don't think fixing it will be any less of a performance loss.

Yes it always registers a new one . Though I don't think it would cause a performence issue . We could optimize some memory by checking if the call back already exist but that will turn it into O(n) from O(1) . Also unless some one has millions of callbacks I don't think memory foot print of lua functions are large enough to make it noticable for anyone .

**Pull request merged**